crates/ruma-client-api/src/sync/sync_events/v4.rs

```diff
@@ -0,0 +1,343 @@
+pub struct SyncList {
+    /// The sync operation to apply, if any.
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    pub ops: Vec<SyncOp>,
+
+    /// The total number of rooms found for this filter.
+    pub count: UInt,
+}
+
+/// Updates to joined rooms.
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub struct SyncOp {
+    /// The sync operation to apply.
+    pub op: SlidingOp,
+
+    /// The range this list update applies to.
+    pub range: Option<(UInt, UInt)>,
+
+    /// Or the specific index the update applies to.
+    pub index: Option<UInt>,
+
+    /// The list of room_ids updates to apply.
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    pub room_ids: Vec<OwnedRoomId>,
+
+    /// On insert and delete we are only receiving exactly one room_id.
+    pub room_id: Option<OwnedRoomId>,
+}
+
+/// Updates to joined rooms.
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub struct SlidingSyncRoom {
+    /// The name of the room as calculated by the server.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub name: Option<String>,
+
+    /// Was this an initial response.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub initial: Option<bool>,
+
+    /// This is a direct message.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub is_dm: Option<bool>,
+
+    /// This is not-yet-accepted invite, with the following sync state events
+    /// the room must be considered in invite state as long as the Option is not None
+    /// even if there are no state events.
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    pub invite_state: Vec<Raw<AnyStrippedStateEvent>>,
+
+    /// Counts of unread notifications for this room.
+    #[serde(flatten, default, skip_serializing_if = "UnreadNotificationsCount::is_empty")]
+    pub unread_notifications: UnreadNotificationsCount,
+
+    /// The timeline of messages and state changes in the room.
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    pub timeline: Vec<Raw<AnySyncRoomEvent>>,
+
+    /// Updates to the state at the beginning of the `timeline`.
+    /// A list of state events.
+    #[serde(default, skip_serializing_if = "Vec::is_empty")]
+    pub required_state: Vec<Raw<AnySyncStateEvent>>,
+
+    /// The prev_batch allowing you to paginate through the messages before the given ones.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub prev_batch: Option<String>,
+}
+
+impl SlidingSyncRoom {
+    /// Creates an empty `Room`.
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Returns true if there are no updates in the room.
+    pub fn is_empty(&self) -> bool {
+        self.unread_notifications.is_empty()
+            && self.timeline.is_empty()
+            && self.required_state.is_empty()
+    }
```

Comment:
What is this for? For other types we have `is_empty` to decide whether to skip serializing a field, that doesn't seem to apply here.